# Lab 1 Machine Learning

### Authors:
Alfredo Ibias Martínez  
Miguel Benito Parejo

The aim of this Lab is to preprocess and audit a given dataset. In order to do so, we analyse the data to prepare it for future uses on machine learning tasks.  
The given dataset has 32561 entries with the following variables:  
- age
- workclass
- fnlwgt
- education
- education-num
- marital-status
- ocupation
- relationship
- race
- sex
- capital-gain
- capital-loss
- hours-per-week
- native-country

And they are classified in the following two classes:
- <=50K
- \>50K

First of all, we need to import the libraries we will use along the lab.

In [1]:
#Libraries

#import numpy as np # linear algebra
import pandas as pd # data processing, CSV file I/O (e.g. pd.read_csv)
import sklearn as skl
from sklearn import preprocessing
import matplotlib.pyplot as plt

Now, we can upload our dataset.

In [2]:
# Read Dataset

dataset = pd.read_csv('./Adult_Income.txt',delimiter=',',header=0)
#dataset

# Header
header = []
for row in dataset:
    header.append(row)
#header

Once we have upload our dataset, we face our first problem. It was what to do with unknown data.  
As our database is large, we first decided to check how many entries lack some data. Therefore, we delete these entries from our dataset and check how many entries we still have.

###### 1  
First, lets see what we have.

In [3]:
print(dataset.describe())

                age        fnlwgt   education-num   capital-gain  \
count  32561.000000  3.256100e+04    32561.000000   32561.000000   
mean      38.581647  1.897784e+05       10.080679    1077.648844   
std       13.640433  1.055500e+05        2.572720    7385.292085   
min       17.000000  1.228500e+04        1.000000       0.000000   
25%       28.000000  1.178270e+05        9.000000       0.000000   
50%       37.000000  1.783560e+05       10.000000       0.000000   
75%       48.000000  2.370510e+05       12.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week         class  
count   32561.000000     32561.000000  32561.000000  
mean       87.303830        40.437456      0.240810  
std       402.960219        12.347429      0.427581  
min         0.000000         1.000000      0.000000  
25%         0.000000        40.000000      0.000000  
50%         0.000000        40.000000      0.000000  
75%      

In [4]:
pos = []
for column in dataset.columns:
    count = 0
    for elem in dataset[column]:
        if elem == " ?":
            dataset = dataset.drop(count)
            pos.append(count)
        count += 1
        while pos.__contains__(count):
            count += 1

In [5]:
print(dataset.describe())

                age        fnlwgt   education-num   capital-gain  \
count  30162.000000  3.016200e+04    30162.000000   30162.000000   
mean      38.437902  1.897938e+05       10.121312    1092.007858   
std       13.134665  1.056530e+05        2.549995    7406.346497   
min       17.000000  1.376900e+04        1.000000       0.000000   
25%       28.000000  1.176272e+05        9.000000       0.000000   
50%       37.000000  1.784250e+05       10.000000       0.000000   
75%       47.000000  2.376285e+05       13.000000       0.000000   
max       90.000000  1.484705e+06       16.000000   99999.000000   

        capital-loss   hours-per-week         class  
count   30162.000000     30162.000000  30162.000000  
mean       88.372489        40.931238      0.248922  
std       404.298370        11.979984      0.432396  
min         0.000000         1.000000      0.000000  
25%         0.000000        40.000000      0.000000  
50%         0.000000        40.000000      0.000000  
75%      

As we can see, before deleting any entry we had 32561 entries, and now we left 30162 entries.

In [6]:
initial = 32561
final = 30162
deleted = initial-final
deleted

2399

In [7]:
percentage = deleted/initial
percentage

0.07367709836921471

Therefore, we have deleted 2399 entries, that are only a 7.37% of the total entries. Thus, we can assume that loss. Also, we have analysed the distribution of deleted entries and he have concluded that they are not significative enough.

Now, lets focus on the education and education-num variables. Doing a quick analysis we have shown that they are exactly the same variable, but expressed in a different data type, so we decided to delete the education column. Also, according to our criteria and most of the information we found on the Internet, we found that the column fnlwgt is an almost random number. Therefore, we decided to delete it.

In [8]:
dataset = dataset.drop(" fnlwgt", axis=1)
dataset = dataset.drop(" education", axis=1)

Once we have deleted the extra or incomplete information, we can focus on nominal variables. In our case, we decided to transform them into a set of dummy variables, so we can work with them properly, in the sense that we don't give any one possible value more relevance than others.

In [9]:
dataset = pd.get_dummies(dataset)

The last transformation we will do to our dataset will be the detection and management of outliers. In order to do so, first, we need to compute the boundaries.

In [10]:
iqr = dataset.quantile(q=0.75, axis=0)-dataset.quantile(q=0.25, axis=0)
q1 = dataset.quantile(q=0.25, axis=0) - 1.5*iqr
q3 = dataset.quantile(q=0.75, axis=0) + 1.5*iqr

Along all the possible outliers, we decided that the only ones that we think we can delete are the ones from the variable education-num, because we think the others are relevant enough to keep them as they follow a wider distribution.

In [11]:
i = 0
poss = []
for elem in dataset[header[4]] < q1[" education-num"]: #They are 196
    if elem:
        poss.append(i)
    i += 1
    while pos.__contains__(i):
        i += 1

for i in poss:
    dataset = dataset.drop(i)

Finally, we have our database ready for work.

Now, lets discuss the points from the instructions:

### Descripción de las variables y valores estadísticos (minimo, máximo, media, desviación,mediana, etc.). Estudia qué valores estadísticos son los convenientes según el tipo de variable y procede en consecuencia.

The variable types are:
- age (numerical)
- workclass (nominal) --> To dummy variables
- fnlwgt (numerical) --> Deleted
- education (nominal) --> Deleted
- education-num (numerical)
- marital-status  (nominal) --> To dummy variables
- ocupation  (nominal) --> To dummy variables
- relationship  (nominal) --> To dummy variables
- race (nominal) --> To dummy variables
- sex (nominal) --> To dummy variables
- capital-gain (numerical)
- capital-loss (numerical)
- hours-per-week (numerical)
- native-country (nominal) --> To dummy variables

The statistical values have been shown at the beginning, at 1. As we can see, there's no statistical values for nominal variables, as they cannot be numerically analysed.

### Describe y realiza modificaciones en la base datos si lo consideras necesario. Por ejemplo, qué harías con valores nominales, si los hubiera.

We have transformed the database previously. We did it by:
- Deleting entries with missing values.
- Deleting the columns of education and fnlwgt.
- Transforming nominal variables into dummy variables.
- Detecting and treating outliers.

### Estudia si es necesario normalizar los datos y cómo lo harías. Procede a modificar la base de datos (normalizar) si lo consideras necesario.

We analysed the database and concluded that we don't need to normalise any variable, as it wouldn't give us any extra information.

### Detección de valores extremos (outliers) y descripción de qué harías en cada caso.

We have detected them and decided to act as follows:
- Delete the outliers of the variable education-num.
- Keep the outliers of the variables age and hours-per-week because they follow a wider distribution.
- Keep the outliers of the variables capital-gain and capital-loss because they would provide no information otherwise.

### Detección de valores perdidos (missing values) y descripción de cómo actuarías para solventar el problema.

We detected them and we decided that they are few enogh to keep an useful database after deleting them.

### Buscar correlaciones entre:
##### las variables predictoras, lo que permitirá ver si hay variables redundantes.
We only found a high correlation between the variables education and education-num, and therefore we deleted the variable education. Also, we found a high correlation between the variables marital-status_ Married-civ-spouse and relationship_ Husband, of 0.9, but as they are dummy variables, we still keep them.
##### variables predictoras y la clase (target).
We found no correlation between the variables and the target class enough to claim that they are a predictive variable. Anyway, we found a high correlation between the class and the variable marital-status_ Married-civ-spouse, of 0.45, but not enough to assert that it is a predictive variable.

### Detecta, si hubiera, falsos predictores.
As we have no variable with a high enough correlation with the target class, we have no false predictive variable to be deleted.

### Estudia si fuera conveniente segmentar alguna de las variables.
We analysed the variables and determined that the only variable that could be segmentated is the age variable.

### Estudia si fuera conveniente crear nuevas variables sintéticas basada en las variables originales.
As we have nominal variables, we determined that it is convenient to create the needed dummy variables.

Note: We used Python3.6 and Spyder in order to see and analyse the data for our work, but these values are not shown in this notebook.